### DISCUSSION TITLE
Typed repositories

### DISCUSSION THREAD

**Discussion by ro0NL** [doctrine/doctrine2#11485] [General]
Would be nice if `$em->getRepository(SomeRepository::class)` worked; to provide actual typed repository return value

**Comment by greg0ire** [comment]
It's already supposed to be the case: https://github.com/doctrine/orm/blob/9d4f54b9a476f13479c3845350b12c466873fc42/src/EntityManagerInterface.php#L25-L31

**Reply by ro0NL** [reply]
no i mean concrete `SomeRepository`, not generic `EntityRepository<SomeEntity>`

**Reply by ro0NL** [reply]
TLDR; SA tools are unaware of custom repository methods

**Reply by derrabus** [reply]
There are plugins for Psalm and PHPStan that solve this issue. I don't see a way to solve this on our side, sorry.

**Reply by ro0NL** [reply]
assuming you actually considered implementing `getRepository($entityOrRepositoryClassname)` it can be typed as such: `@psalm-return (T is EntityRepository ? T : EntityRepository<T>)`

**Reply by derrabus** [reply]
But that call would be ambiguous. We e.g. wouldn't know for which entity we need to load that repository. Also you would need to rewrite your code.

Just use those plugins. They do the job.
